exp/controllers/awsmachinepool_controller.go

```diff
@@ -392,6 +425,163 @@ func (r *AWSMachinePoolReconciler) reconcileDelete(machinePoolScope *scope.Machi
+				},
+				Spec: infrav1.AWSMachineSpec{
+					ProviderID: aws.String(providerID),
+					InstanceID: aws.String(instanceID),
+
+					// (cmcavoy): XXX not sure if it's useful or misleading to convert ec2 fields over to the AWSMachine
+					AMI: infrav1.AMIReference{
+						ID: aws.String(instance.ImageID),
+					},
+					InstanceType:             instance.Type,
+					PublicIP:                 aws.Bool(instance.PublicIP != nil),
+					SSHKeyName:               instance.SSHKeyName,
+					InstanceMetadataOptions:  instance.InstanceMetadataOptions,
+					IAMInstanceProfile:       instance.IAMProfile,
+					AdditionalSecurityGroups: securityGroups,
+					Subnet:                   &infrav1.AWSResourceReference{ID: aws.String(instance.SubnetID)},
+					RootVolume:               instance.RootVolume,
+					NonRootVolumes:           instance.NonRootVolumes,
+					NetworkInterfaces:        instance.NetworkInterfaces,
+					CloudInit:                infrav1.CloudInit{},
+					SpotMarketOptions:        instance.SpotMarketOptions,
+					Tenancy:                  instance.Tenancy,
+				},
+			}
+			instanceLogger.V(2).Info("creating AWSMachine instance")
+			if err := client.Create(ctx, awsMachine); err != nil {
+				return fmt.Errorf("failed to create AWSMachine: %w", err)
+			}
+		}
+	}
+	return nil
+}
+
+func deleteOrphanedAWSMachines(ctx context.Context, awsMachineList *infrav1.AWSMachineList, existingASG *expinfrav1.AutoScalingGroup, client client.Client) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("asg", existingASG.Name)
+	log.V(2).Info("Delete orphaned awsmachines")
+
+	providerIDs := make(map[string]infrav1.Instance, len(existingASG.Instances))
+	for i := range existingASG.Instances {
+		providerID := fmt.Sprintf("aws:///%s/%s", existingASG.Instances[i].AvailabilityZone, existingASG.Instances[i].ID)
+		providerIDs[providerID] = existingASG.Instances[i]
+	}
+
+	for i := range awsMachineList.Items {
+		awsMachine := awsMachineList.Items[i]
+		if awsMachine.Spec.ProviderID == nil || *awsMachine.Spec.ProviderID == "" {
+			continue
+		}
+
+		providerID := *awsMachine.Spec.ProviderID
+		if _, exists := providerIDs[providerID]; !exists {
+			machine, err := util.GetOwnerMachine(ctx, client, awsMachine.ObjectMeta)
+			if err != nil {
+				return fmt.Errorf("failed to get owner machine for %s/%s: %w", awsMachine.Namespace, awsMachine.Name, err)
+			}
+			log.V(2).Info("Deleting orphaned machine", "machine", machine, "awsmachine", awsMachine, "ProviderID", providerID)
+			if machine == nil {
+				// XXX(cmcavoy): if we got here, something went wrong with the owner reference
+				if err := client.Delete(ctx, &awsMachine); err != nil {
+					return fmt.Errorf("failed to delete orphan awsMachine %s/%s: %w", awsMachine.Namespace, awsMachine.Name, err)
+				}
+				continue
+			}
+
+			if err := client.Delete(ctx, machine); err != nil {
+				return fmt.Errorf("failed to delete orphan machine %s/%s: %w", machine.Namespace, machine.Name, err)
+			}
```

Comment:
So if I understand this correctly... 

If there is a machine for which there is no provider... check if there is an owner. If there is an owner, we delete the owner? And then we hope that cascading delete will also delete this machine?

Reply:
That is how I parsed the proposal language:

> When a MachinePool Machine is deleted manually, the system will delete the corresponding provider-specific resource. The opposite is also true: when a provider-specific resource is deleted, the system will delete the corresponding MachinePool Machine. This happens by virtue of the infrastructureRef <-> ownerRef relationship.

https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220209-machinepool-machines.md#proposal

@jont828 @mboersma @devigned wrote the proposal and perhaps can provide clarity ...?

What is the expected outcome when an AWSMachine and Machine pair, which is owned by a MachinePool, when it's ec2 instance is destroyed via an external action (e.g user terminates the instance in the aws console). Should the infrastructure provider (CAPA) detect this termination and delete the CAPI Machine resource?

Reply:
Yes, if an AWSMachine's associated provider instance is deleted, we expect the AWSMachinePool to delete the Machine and AWSMachine pair. This is so that we are not left with hanging resources in the event of an out-of-band instance deletion like you described.

Reply:
Yes, but... if we delete its owner, is it okay that any OTHER machine this thing owns will also get deleted?

Reply:
I'm not quite sure I understand. Are you saying that if you delete the CAPI Machine, i.e. owner of the AWS Machine, you want to also delete another CAPI Machine associated with the same instance?

Reply:
No, I mean is it a problem if the deleted owner owns more machines? Those would also be deleted then.

Reply:
When you delete the CAPI Machine, we expect it to delete the AWSMachine. It would be the same behavior we already have with MachineDeployments except there is no KubeadmConfig object to delete.